lusSTR/filter_settings.py

```diff
@@ -321,6 +349,53 @@ def allele_counts(allele_df):
     return mix_df
 
 
+def create_repeat_dict(bracket):
+    repeats = defaultdict(lambda: defaultdict(dict))
+    num = 0
+    for unit in bracket.split(" "):
+        num += 1
+        if "[" in unit:
+            match = re.match(r"\[([ACGT]+)\](\d+)", unit)
+            if match:
+                repeat_motif = match.group(1)
+                count = match.group(2)
+        else:
+            repeat_motif = unit
+            count = 1
+        repeats[num]["motif"] = repeat_motif
+        repeats[num]["counts"] = int(count)
+    return repeats
+
+
+def bracketed_stutter_id(ref_bracket, quest_bracket, stutter_id):
+    ref_repeats = create_repeat_dict(ref_bracket)
+    quest_repeats = create_repeat_dict(quest_bracket)
+    stutter = None
+    if len(quest_repeats) == len(ref_repeats):
+        diffcount = 0
+        for key, value in ref_repeats.items():
+            ref_info = ref_repeats[key]
+            quest_info = quest_repeats[key]
+            if ref_info["motif"] != quest_info["motif"]:
+                stutter = None
+                break
+            else:
+                diff = quest_info["counts"] - ref_info["counts"]
+                if diff == 0:
+                    continue
+                elif diff == stutter_id:
+                    if diffcount == 0:
+                        diffcount = diff
+                        stutter = stutter_id
+                    else:
+                        stutter = None
+                        break
+                else:
+                    stutter = None
+                    break
+    return stutter
```

Comment:
These are the two main functions for calling stutter using the sequence bracketed form. The program first determines the potential type of stutter (e.g. -1, -2, +1) using the CE allele and then feeds the bracketed form into these functions. It reads in the brackets into a dict for each sequence (one reference and one questioned) and compares each repeat motif to compare the # of repeats and ensures only one of the repeat motifs matches the type of stutter (e.g. ```[AGAT]4 [ACAT]7``` and ```[AGAT]3 [ACAT]7 ``` is the -1 stutter sequence but ```[AGAT]2[ACAT]8``` is not even though they'd both have the same CE allele. It's actually fairly simple code for something that seemed so complex in theory...

Reply:
These two functions are pretty clear. It looks like the code assumes that the `ref_repeats` and `quest_repeats` dictionaries will have the same keys. This is because we only do the check if they have the same *number* of keys, and they keys are just indices, right?

Reply:
Yes, the keys are just indices. While the first thing the code does is make sure the number of keys are identical between the reference and questioned (because that's the easiest way to identify differences right off the bat), it does loop through the indices to check that the motif of that particular index does match between the reference and questioned sets (the ```ifelse``` statement starting on line 379).

Reply:
I think that answers your question. We can talk in person on Monday if you want further clarification.